File: app/src/main/java/com/doubletap/screenoff/MainActivity.java
package com.doubletap.screenoff;

import android.accessibilityservice.AccessibilityService;
import android.app.WallpaperManager;
import android.app.admin.DevicePolicyManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.media.AudioManager;
import android.media.ToneGenerator;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.PowerManager;
import android.os.Vibrator;
import android.provider.Settings;
import android.text.TextUtils;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.card.MaterialCardView;
import com.google.android.material.materialswitch.MaterialSwitch;

public class MainActivity extends AppCompatActivity {

    private TextView tvAccessibilityStatus;
    private TextView tvAdminStatus;
    private TextView tvWidgetStatus;
    private TextView tvSensorStatus;

    private MaterialSwitch switchFloatingWidget;
    private MaterialSwitch switchAccelerometer;
    private MaterialSwitch switchSound;
    private MaterialSwitch switchVibration;

    private Button btnToggleAccessibility;
    private Button btnToggleAdmin;
    private Button btnIgnoreBattery;
    private Button btnChooseWallpaper;
    private Button btnSetLiveWallpaper;
    private Button btnResetWallpaper;
    private MaterialCardView btnTestPad;
    private TextView tvAbout;
    private TextView tvPrivacy;

    private SharedPreferences prefs;
    private static final int PICK_IMAGE_REQUEST = 1001;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        prefs = getSharedPreferences("doubletap_prefs", MODE_PRIVATE);

        // Bind Views
        tvAccessibilityStatus = findViewById(R.id.tvAccessibilityStatus);
        tvAdminStatus = findViewById(R.id.tvAdminStatus);
        tvWidgetStatus = findViewById(R.id.tvWidgetStatus);
        tvSensorStatus = findViewById(R.id.tvSensorStatus);

        switchFloatingWidget = findViewById(R.id.switchFloatingWidget);
        switchAccelerometer = findViewById(R.id.switchAccelerometer);
        switchSound = findViewById(R.id.switchSound);
        switchVibration = findViewById(R.id.switchVibration);

        btnToggleAccessibility = findViewById(R.id.btnToggleAccessibility);
        btnToggleAdmin = findViewById(R.id.btnToggleAdmin);
        btnIgnoreBattery = findViewById(R.id.btnIgnoreBattery);
        btnChooseWallpaper = findViewById(R.id.btnChooseWallpaper);
        btnSetLiveWallpaper = findViewById(R.id.btnSetLiveWallpaper);
        btnResetWallpaper = findViewById(R.id.btnResetWallpaper);
        btnTestPad = findViewById(R.id.btnTestPad);
        tvAbout = findViewById(R.id.tvAbout);
        tvPrivacy = findViewById(R.id.tvPrivacy);

        // Load saved preferences
        switchFloatingWidget.setChecked(prefs.getBoolean("enable_floating_widget", false));
        switchAccelerometer.setChecked(prefs.getBoolean("enable_accelerometer", false));
        switchSound.setChecked(prefs.getBoolean("enable_sound", true));
        switchVibration.setChecked(prefs.getBoolean("enable_vibration", true));

        // Click listeners
        btnToggleAccessibility.setOnClickListener(v -> {
            Intent intent = new Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS);
            startActivity(intent);
            Toast.makeText(MainActivity.this, "Find 'Double Tap Screen Off' and enable it", Toast.LENGTH_LONG).show();
        });

        btnToggleAdmin.setOnClickListener(v -> {
            DevicePolicyManager dpm = (DevicePolicyManager) getSystemService(Context.DEVICE_POLICY_SERVICE);
            ComponentName adminName = new ComponentName(MainActivity.this, ScreenLockAdminReceiver.class);
            if (dpm != null && !dpm.isAdminActive(adminName)) {
                Intent intent = new Intent(DevicePolicyManager.ACTION_ADD_DEVICE_ADMIN);
                intent.putExtra(DevicePolicyManager.EXTRA_DEVICE_ADMIN, adminName);
                intent.putExtra(DevicePolicyManager.EXTRA_ADD_EXPLANATION, getString(R.string.device_admin_description));
                startActivity(intent);
            } else if (dpm != null) {
                dpm.removeActiveAdmin(adminName);
                Toast.makeText(MainActivity.this, "Device Administrator Fallback deactivated", Toast.LENGTH_SHORT).show();
                updateStatuses();
            }
        });

        btnIgnoreBattery.setOnClickListener(v -> {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                Intent intent = new Intent();
                String packageName = getPackageName();
                PowerManager pm = (PowerManager) getSystemService(POWER_SERVICE);
                if (pm != null && !pm.isIgnoringBatteryOptimizations(packageName)) {
                    intent.setAction(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS);
                    intent.setData(Uri.parse("package:" + packageName));
                    startActivity(intent);
                } else {
                    Toast.makeText(MainActivity.this, "App is already whitelisted from battery optimization", Toast.LENGTH_SHORT).show();
                }
            } else {
                Toast.makeText(MainActivity.this, "Battery optimization not restricted on this SDK", Toast.LENGTH_SHORT).show();
            }
        });

        // Toggle handling
        switchFloatingWidget.setOnCheckedChangeListener((buttonView, isChecked) -> {
            prefs.edit().putBoolean("enable_floating_widget", isChecked).apply();
            if (isChecked) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(MainActivity.this)) {
                    buttonView.setChecked(false);
                    Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse("package:" + getPackageName()));
                    startActivity(intent);
                    Toast.makeText(MainActivity.this, "Please grant overlay permission", Toast.LENGTH_LONG).show();
                } else {
                    startFloatingWidgetService();
                }
            } else {
                stopFloatingWidgetService();
            }
            updateStatuses();
        });

        switchAccelerometer.setOnCheckedChangeListener((buttonView, isChecked) -> {
            prefs.edit().putBoolean("enable_accelerometer", isChecked).apply();
            if (isChecked) {
                startSensorService();
            } else {
                stopSensorService();
            }
            updateStatuses();
        });

        switchSound.setOnCheckedChangeListener((buttonView, isChecked) -> {
            prefs.edit().putBoolean("enable_sound", isChecked).apply();
        });

        switchVibration.setOnCheckedChangeListener((buttonView, isChecked) -> {
            prefs.edit().putBoolean("enable_vibration", isChecked).apply();
        });

        // Test lock gesture on main pad
        GestureDetector gestureDetector = new GestureDetector(this, new GestureDetector.SimpleOnGestureListener() {
            @Override
            public boolean onDoubleTap(MotionEvent e) {
                performLockRoutine(true);
                return true;
            }
        });

        btnTestPad.setOnTouchListener((v, event) -> {
            gestureDetector.onTouchEvent(event);
            return true;
        });

        tvAbout.setOnClickListener(v -> showAboutDialog());
        tvPrivacy.setOnClickListener(v -> showPrivacyDialog());

        // Wallpaper Customization Click Listeners
        btnChooseWallpaper.setOnClickListener(v -> {
            Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
            intent.setType("image/*");
            startActivityForResult(Intent.createChooser(intent, "Select Wallpaper Photo"), PICK_IMAGE_REQUEST);
        });

        btnSetLiveWallpaper.setOnClickListener(v -> {
            try {
                Intent intent = new Intent(WallpaperManager.ACTION_CHANGE_LIVE_WALLPAPER);
                intent.putExtra(WallpaperManager.EXTRA_LIVE_WALLPAPER_COMPONENT,
                        new ComponentName(MainActivity.this, DoubleTapWallpaperService.class));
                startActivity(intent);
            } catch (Exception e) {
                Toast.makeText(MainActivity.this, "Your device launcher may require manual setting from Home Screen -> Wallpapers.", Toast.LENGTH_LONG).show();
            }
        });

        btnResetWallpaper.setOnClickListener(v -> {
            java.io.File file = new java.io.File(getFilesDir(), "custom_wallpaper.png");
            if (file.exists()) {
                file.delete();
            }
            Toast.makeText(MainActivity.this, "Wallpaper reset to default slate theme color!", Toast.LENGTH_SHORT).show();
        });

        // Check for overlay permission fallback
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && switchFloatingWidget.isChecked() && Settings.canDrawOverlays(this)) {
            startFloatingWidgetService();
        }

        // Ask the user to enable Accessibility Service on launch if not active
        if (!isAccessibilityServiceEnabled(this, DoubleTapAccessibilityService.class)) {
            showAccessibilityPromptDialog();
        }
    }

    private void showAccessibilityPromptDialog() {
        new AlertDialog.Builder(this, androidx.appcompat.R.style.Theme_AppCompat_Dialog_Alert)
                .setTitle("Enable Accessibility Service")
                .setMessage("Double Tap Screen Off requires the Accessibility Service permission to perform secure screen locking without physical buttons. This service is lightweight, fully local, offline, and safe. Please tap 'Enable' to open settings and select Double Tap Screen Off.")
                .setCancelable(false)
                .setPositiveButton("Enable", (dialog, which) -> {
                    Intent intent = new Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS);
                    startActivity(intent);
                    Toast.makeText(this, "Find 'Double Tap Screen Off' under Downloaded Services and enable it", Toast.LENGTH_LONG).show();
                })
                .setNegativeButton("Later", null)
                .show();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == PICK_IMAGE_REQUEST && resultCode == RESULT_OK && data != null && data.getData() != null) {
            android.net.Uri imageUri = data.getData();
            try {
                java.io.InputStream inputStream = getContentResolver().openInputStream(imageUri);
                if (inputStream != null) {
                    java.io.File file = new java.io.File(getFilesDir(), "custom_wallpaper.png");
                    java.io.FileOutputStream outputStream = new java.io.FileOutputStream(file);
                    
                    byte[] buffer = new byte[4096];
                    int bytesRead;
                    while ((bytesRead = inputStream.read(buffer)) != -1) {
                        outputStream.write(buffer, 0, bytesRead);
                    }
                    
                    inputStream.close();
                    outputStream.close();
                    
                    Toast.makeText(this, "Wallpaper successfully updated! Apply the Live Wallpaper to see changes.", Toast.LENGTH_LONG).show();
                }
            } catch (Exception e) {
                Toast.makeText(this, "Failed to load wallpaper: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                e.printStackTrace();
            }
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        updateStatuses();
    }

    private void updateStatuses() {
        // 1. Accessibility Service Enabled
        boolean accEnabled = isAccessibilityServiceEnabled(this, DoubleTapAccessibilityService.class);
        if (accEnabled) {
            tvAccessibilityStatus.setText(R.string.service_active_status);
            tvAccessibilityStatus.setTextColor(getResources().getColor(R.color.status_green));
        } else {
            tvAccessibilityStatus.setText(R.string.service_inactive_status);
            tvAccessibilityStatus.setTextColor(getResources().getColor(R.color.status_red));
        }

        // 2. Device Admin Active
        DevicePolicyManager dpm = (DevicePolicyManager) getSystemService(Context.DEVICE_POLICY_SERVICE);
        ComponentName adminName = new ComponentName(this, ScreenLockAdminReceiver.class);
        boolean adminActive = dpm != null && dpm.isAdminActive(adminName);
        if (adminActive) {
            tvAdminStatus.setText(R.string.service_active_status);
            tvAdminStatus.setTextColor(getResources().getColor(R.color.status_green));
            btnToggleAdmin.setText("Disable Device Admin Fallback");
        } else {
            tvAdminStatus.setText(R.string.service_inactive_status);
            tvAdminStatus.setTextColor(getResources().getColor(R.color.status_red));
            btnToggleAdmin.setText("Enable Device Admin Fallback");
        }

        // 3. Floating Widget service status
        boolean widgetRunning = FloatingWidgetService.isRunning();
        if (widgetRunning) {
            tvWidgetStatus.setText(R.string.service_active_status);
            tvWidgetStatus.setTextColor(getResources().getColor(R.color.status_green));
        } else {
            tvWidgetStatus.setText(R.string.service_inactive_status);
            tvWidgetStatus.setTextColor(getResources().getColor(R.color.status_red));
        }

        // 4. Sensor background service status
        boolean sensorRunning = DoubleTapSensorService.isRunning();
        if (sensorRunning) {
            tvSensorStatus.setText(R.string.service_active_status);
            tvSensorStatus.setTextColor(getResources().getColor(R.color.status_green));
        } else {
            tvSensorStatus.setText(R.string.service_inactive_status);
            tvSensorStatus.setTextColor(getResources().getColor(R.color.status_red));
        }
    }

    private void startFloatingWidgetService() {
        Intent serviceIntent = new Intent(this, FloatingWidgetService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(serviceIntent);
        } else {
            startService(serviceIntent);
        }
    }

    private void stopFloatingWidgetService() {
        stopService(new Intent(this, FloatingWidgetService.class));
    }

    private void startSensorService() {
        Intent serviceIntent = new Intent(this, DoubleTapSensorService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(serviceIntent);
        } else {
            startService(serviceIntent);
        }
    }

    private void stopSensorService() {
        stopService(new Intent(this, DoubleTapSensorService.class));
    }

    private void performLockRoutine(boolean isInteractiveTest) {
        // Feedback vibration
        if (prefs.getBoolean("enable_vibration", true)) {
            Vibrator vibrator = (Vibrator) getSystemService(VIBRATOR_SERVICE);
            if (vibrator != null && vibrator.hasVibrator()) {
                vibrator.vibrate(70);
            }
        }

        // Feedback sound
        if (prefs.getBoolean("enable_sound", true)) {
            try {
                ToneGenerator toneGen = new ToneGenerator(AudioManager.STREAM_SYSTEM, 70);
                toneGen.startTone(ToneGenerator.TONE_PROP_PROMPT, 80);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        if (isInteractiveTest) {
            Toast.makeText(this, "Lock Triggered!", Toast.LENGTH_SHORT).show();
        }

        // Broadcast or execute lock action
        Intent lockIntent = new Intent("com.doubletap.screenoff.ACTION_LOCK");
        sendBroadcast(lockIntent);
    }

    public static boolean isAccessibilityServiceEnabled(Context context, Class<? extends AccessibilityService> service) {
        ComponentName expectedComponentName = new ComponentName(context, service);
        String enabledServicesSetting = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.ENABLED_ACCESSIBILITY_SERVICES);
        if (enabledServicesSetting == null) return false;
        TextUtils.SimpleStringSplitter colonSplitter = new TextUtils.SimpleStringSplitter(':');
        colonSplitter.setString(enabledServicesSetting);
        while (colonSplitter.hasNext()) {
            String componentNameString = colonSplitter.next();
            ComponentName enabledService = ComponentName.unflattenFromString(componentNameString);
            if (enabledService != null && enabledService.equals(expectedComponentName)) return true;
        }
        return false;
    }

    private void showAboutDialog() {
        new AlertDialog.Builder(this, androidx.appcompat.R.style.Theme_AppCompat_Dialog_Alert)
                .setTitle("About Double Tap Screen Off")
                .setMessage("Double Tap Screen Off is a lightweight utility designed to help users quickly lock their mobile device without clicking physical buttons. By leveraging a high-performance background sensor service and secure Accessibility framework, it provides immediate reaction speeds with less than 0.1% battery consumption daily.\n\nDeveloped entirely in native Java.")
                .setPositiveButton("OK", null)
                .show();
    }

    private void showPrivacyDialog() {
        new AlertDialog.Builder(this, androidx.appcompat.R.style.Theme_AppCompat_Dialog_Alert)
                .setTitle("Privacy Policy")
                .setMessage("Your privacy is absolute. Double Tap Screen Off:\n\n• Does NOT request Internet permissions.\n• Does NOT track or collect user analytics or telemetry.\n• Operates entirely on-device offline.\n• Standard Android security permissions are requested strictly to perform the screen-locking action on double tap.")
                .setPositiveButton("OK", null)
                .show();
    }
}

File: app/src/main/java/com/doubletap/screenoff/MainActivity.kt
package com.doubletap.screenoff

import android.app.WallpaperManager
import android.app.admin.DevicePolicyManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

class MainActivity : ComponentActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            DoubleTapLockTheme {
                DashboardScreen()
            }
        }
    }

    @Composable
    fun DoubleTapLockTheme(content: @Composable () -> Unit) {
        MaterialTheme(
            colorScheme = darkColorScheme(
                primary = Color(0xFF818CF8), // Indigo Light
                secondary = Color(0xFFA5B4FC),
                background = Color(0xFF0F172A), // Dark Navy Slate
                surface = Color(0xFF1E293B)
            ),
            content = content
        )
    }

    @OptIn(ExperimentalMaterial3Api::class)
    @Composable
    fun DashboardScreen() {
        val context = LocalContext.current
        val scrollState = rememberScrollState()

        // State Checkers
        var isAccessibilityOn by remember { mutableStateOf(false) }
        var isDeviceAdminOn by remember { mutableStateOf(false) }
        var isOverlayAllowed by remember { mutableStateOf(false) }
        var isFloatingServiceRunning by remember { mutableStateOf(false) }

        // Periodic Check on lifecycle resume
        val checkStates = {
            isAccessibilityOn = isAccessibilityServiceEnabled(context, AccessibilityLockService::class.java)
            isDeviceAdminOn = isDeviceAdminEnabled(context)
            isOverlayAllowed = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                Settings.canDrawOverlays(context)
            } else {
                true
            }
        }

        LaunchedEffect(Unit) {
            checkStates()
        }

        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(
                    Brush.verticalGradient(
                        colors = listOf(
                            Color(0xFF0F172A), // Slate 900
                            Color(0xFF312E81), // Indigo 900
                            Color(0xFF581C87)  // Purple 900
                        )
                    )
                )
                .padding(20.dp)
        ) {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .verticalScroll(scrollState),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Spacer(modifier = Modifier.height(20.dp))

                // Release Status Tag Badge
                Box(
                    modifier = Modifier
                        .clip(CircleShape)
                        .background(Color.White.copy(alpha = 0.1f))
                        .padding(horizontal = 14.dp, vertical = 6.dp)
                ) {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        Box(
                            modifier = Modifier
                                .size(8.dp)
                                .clip(CircleShape)
                                .background(if (isAccessibilityOn) Color(0xFF34D399) else Color(0xFFEF4444))
                        )
                        Spacer(modifier = Modifier.width(8.dp))
                        Text(
                            text = "V1.0.5 • ${if (isAccessibilityOn) "SERVICE STABLE" else "SETUP REQUIRED"}",
                            color = Color(0xFFC7D2FE),
                            fontSize = 11.sp,
                            fontWeight = FontWeight.Bold
                        )
                    }
                }

                Spacer(modifier = Modifier.height(24.dp))

                // Title
                Text(
                    text = "Double Tap\nScreen Off",
                    color = Color.White,
                    fontSize = 38.sp,
                    fontWeight = FontWeight.ExtraBold,
                    lineHeight = 44.sp,
                    textAlign = TextAlign.Center
                )

                Spacer(modifier = Modifier.height(12.dp))

                // Subtitle
                Text(
                    text = "Programmatic screen power-off utility for Android. Fully battery-safe daemon.",
                    color = Color.White.copy(alpha = 0.7f),
                    fontSize = 14.sp,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.padding(horizontal = 16.dp)
                )

                Spacer(modifier = Modifier.height(30.dp))

                // Telemetry Card
                Card(
                    shape = RoundedCornerShape(24.dp),
                    colors = CardDefaults.cardColors(containerColor = Color.White.copy(alpha = 0.06f)),
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Column(modifier = Modifier.padding(20.dp)) {
                        Text(
                            text = "SYSTEM METRICS",
                            color = Color(0xFFA5B4FC),
                            fontSize = 12.sp,
                            fontWeight = FontWeight.Bold,
                            letterSpacing = 1.5.sp
                        )
                        Spacer(modifier = Modifier.height(16.dp))
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.SpaceBetween
                        ) {
                            Column {
                                Text(text = "BATTERY / DAY", color = Color.White.copy(alpha = 0.5f), fontSize = 11.sp)
                                Text(text = "0.02%", color = Color.White, fontSize = 22.sp, fontWeight = FontWeight.Light)
                            }
                            Column {
                                Text(text = "RAM IMPACT", color = Color.White.copy(alpha = 0.5f), fontSize = 11.sp)
                                Text(text = "14 MB", color = Color.White, fontSize = 22.sp, fontWeight = FontWeight.Light)
                            }
                        }
                    }
                }

                Spacer(modifier = Modifier.height(24.dp))

                Text(
                    text = "REQUIRED CONFIGURATION",
                    color = Color(0xFFA5B4FC),
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Bold,
                    letterSpacing = 1.2.sp,
                    modifier = Modifier
                        .align(Alignment.Start)
                        .padding(start = 4.dp, bottom = 12.dp)
                )

                // 1. Accessibility Toggle Row
                ConfigRow(
                    title = "Accessibility Service",
                    description = "Required to lock screen instantly on modern systems.",
                    statusText = if (isAccessibilityOn) "Active" else "Grant Access",
                    statusColor = if (isAccessibilityOn) Color(0xFF34D399) else Color(0xFFFBBF24),
                    icon = Icons.Default.AccessibilityNew,
                    onClick = {
                        val intent = Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS)
                        context.startActivity(intent)
                        Toast.makeText(context, "Enable 'Double Tap Screen Off' under Downloaded Services", Toast.LENGTH_LONG).show()
                    }
                )

                Spacer(modifier = Modifier.height(12.dp))

                // 2. Set Wallpaper Button Row
                ConfigRow(
                    title = "Home Double-Tap Wallpaper",
                    description = "Sets background live wallpaper listener on home viewport.",
                    statusText = "Setup Wallpaper",
                    statusColor = Color(0xFF818CF8),
                    icon = Icons.Default.Wallpaper,
                    onClick = {
                        try {
                            val intent = Intent(WallpaperManager.ACTION_CHANGE_LIVE_WALLPAPER).apply {
                                putExtra(
                                    WallpaperManager.EXTRA_LIVE_WALLPAPER_COMPONENT,
                                    ComponentName(context, DoubleTapWallpaperService::class.java)
                                )
                            }
                            context.startActivity(intent)
                        } catch (e: Exception) {
                            val intent = Intent(WallpaperManager.ACTION_LIVE_WALLPAPER_CHOOSER)
                            context.startActivity(intent)
                        }
                    }
                )

                Spacer(modifier = Modifier.height(12.dp))

                // 3. Device Admin Privileges Fallback Row
                ConfigRow(
                    title = "Device Administrator",
                    description = "Legacy screen sleep framework fallback interface.",
                    statusText = if (isDeviceAdminOn) "Granted" else "Activate Fallback",
                    statusColor = if (isDeviceAdminOn) Color(0xFF34D399) else Color(0xFFEF4444).copy(alpha = 0.8f),
                    icon = Icons.Default.Security,
                    onClick = {
                        if (!isDeviceAdminOn) {
                            val intent = Intent(DevicePolicyManager.ACTION_ADD_DEVICE_ADMIN).apply {
                                putExtra(DevicePolicyManager.EXTRA_DEVICE_ADMIN, AdminReceiver.getComponentName(context))
                                putExtra(DevicePolicyManager.EXTRA_ADD_EXPLANATION, "Secure locking permissions daemon.")
                            }
                            context.startActivity(intent)
                        } else {
                            val dpm = context.getSystemService(Context.DEVICE_POLICY_SERVICE) as DevicePolicyManager
                            dpm.removeActiveAdmin(AdminReceiver.getComponentName(context))
                            Toast.makeText(context, "Admin Deactivated", Toast.LENGTH_SHORT).show()
                            checkStates()
                        }
                    }
                )

                Spacer(modifier = Modifier.height(12.dp))

                // 4. Floating Shortcut Bubble Row
                ConfigRow(
                    title = "Assistive Floating Bubble",
                    description = "Overlay double-tap shortcut widget.",
                    statusText = if (isFloatingServiceRunning) "Running" else "Spawn Bubble",
                    statusColor = if (isFloatingServiceRunning) Color(0xFF34D399) else Color(0xFFC7D2FE),
                    icon = Icons.Default.TouchApp,
                    onClick = {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(context)) {
                            val intent = Intent(
                                Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                                Uri.parse("package:${context.packageName}")
                            )
                            context.startActivity(intent)
                            Toast.makeText(context, "Allow Draw Over Other Apps permission.", Toast.LENGTH_LONG).show()
                        } else {
                            val serviceIntent = Intent(context, FloatingLockService::class.java)
                            if (isFloatingServiceRunning) {
                                context.stopService(serviceIntent)
                                isFloatingServiceRunning = false
                                Toast.makeText(context, "Floating controller removed.", Toast.LENGTH_SHORT).show()
                            } else {
                                context.startService(serviceIntent)
                                isFloatingServiceRunning = true
                                Toast.makeText(context, "Floating lock bubble spawned!", Toast.LENGTH_SHORT).show()
                            }
                        }
                    }
                )

                Spacer(modifier = Modifier.height(30.dp))

                // Test Programmatic Screen Lock Button
                Button(
                    onClick = {
                        if (isAccessibilityOn) {
                            AccessibilityLockService.instance?.lockScreenDevice()
                        } else if (isDeviceAdminOn) {
                            AdminReceiver.lockWithAdmin(context)
                        } else {
                            Toast.makeText(context, "Please enable Accessibility or Admin first!", Toast.LENGTH_LONG).show()
                        }
                    },
                    colors = ButtonDefaults.buttonColors(containerColor = Color.White, contentColor = Color(0xFF1E1B4B)),
                    shape = RoundedCornerShape(16.dp),
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(58.dp)
                ) {
                    Icon(imageVector = Icons.Default.PlayArrow, contentDescription = null, modifier = Modifier.size(20.dp))
                    Spacer(modifier = Modifier.width(8.dp))
                    Text(text = "TEST PROGRAMMATIC LOCK NOW", fontWeight = FontWeight.Bold, letterSpacing = 0.5.sp)
                }

                Spacer(modifier = Modifier.height(20.dp))
            }
        }
    }

    @Composable
    fun ConfigRow(
        title: String,
        description: String,
        statusText: String,
        statusColor: Color,
        icon: androidx.compose.ui.graphics.vector.ImageVector,
        onClick: () -> Unit
    ) {
        Card(
            shape = RoundedCornerShape(20.dp),
            colors = CardDefaults.cardColors(containerColor = Color.White.copy(alpha = 0.05f)),
            modifier = Modifier
                .fillMaxWidth()
                .clickable { onClick() }
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(16.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Box(
                    modifier = Modifier
                        .size(40.dp)
                        .clip(RoundedCornerShape(12.dp))
                        .background(Color.White.copy(alpha = 0.08f)),
                    contentAlignment = Alignment.Center
                ) {
                    Icon(imageVector = icon, contentDescription = null, tint = Color(0xFFA5B4FC), modifier = Modifier.size(20.dp))
                }

                Spacer(modifier = Modifier.width(16.dp))

                Column(modifier = Modifier.weight(1f)) {
                    Text(text = title, color = Color.White, fontSize = 14.sp, fontWeight = FontWeight.Bold)
                    Text(text = description, color = Color.White.copy(alpha = 0.5f), fontSize = 11.sp, lineHeight = 14.sp)
                }

                Spacer(modifier = Modifier.width(10.dp))

                Box(
                    modifier = Modifier
                        .clip(CircleShape)
                        .background(statusColor.copy(alpha = 0.15f))
                        .padding(horizontal = 10.dp, vertical = 4.dp)
                ) {
                    Text(text = statusText, color = statusColor, fontSize = 10.sp, fontWeight = FontWeight.ExtraBold)
                }
            }
        }
    }

    private fun isAccessibilityServiceEnabled(context: Context, service: Class<*>): Boolean {
        val expectedComponentName = ComponentName(context, service)
        val enabledServicesSetting = Settings.Secure.getString(
            context.contentResolver,
            Settings.Secure.ENABLED_ACCESSIBILITY_SERVICES
        ) ?: return false

        val colonSplitter = android.text.TextUtils.SimpleStringSplitter(':')
        colonSplitter.setString(enabledServicesSetting)
        while (colonSplitter.hasNext()) {
            val componentNameString = colonSplitter.next()
            val enabledService = ComponentName.unflattenFromString(componentNameString)
            if (enabledService != null && enabledService == expectedComponentName) {
                return true
            }
        }
        return false
    }

    private fun isDeviceAdminEnabled(context: Context): Boolean {
        val dpm = context.getSystemService(Context.DEVICE_POLICY_SERVICE) as DevicePolicyManager
        return dpm.isAdminActive(AdminReceiver.getComponentName(context))
    }
}
